packages/dds/merge-tree/src/stamps.ts

```diff
@@ -0,0 +1,157 @@
+ * ordered by their localSeq.
+ *
+ * By including `clientId`, it also provides enough information to resolve whether segments are visible
+ * from alternative perspectives: a remote client will have seen all of its own previous operations as well as
+ * those at or below the op's reference sequence number.
+ *
+ * @remarks - As the `readonly` identifies suggest, these stamps should be treated as immutable.
+ * New operations applied to a merge-tree should create new stamps rather than modify existing ones (e.g. when
+ * a change's ack happens).
+ */
+export interface OperationStamp {
+	/**
+	 * Thge sequence number at which this operation was applied.
+	 */
+	readonly seq: number;
+
+	/**
+	 * Short clientId for the client that performed this operation.
+	 */
+	readonly clientId: number;
+
+	/**
+	 * Local seq at which this operation was applied.
+	 * This is defined if and only if the operation is pending an ack, i.e. `seq` is UnassignedSequenceNumber.
+	 *
+	 * @privateRemarks
+	 * See {@link CollaborationWindow.localSeq} for more information on the semantics of localSeq.
+	 */
+	readonly localSeq?: number;
+}
+
+/**
+ * {@link OperationStamp} for an 'insert' operation.
+ */
+export interface InsertOperationStamp extends OperationStamp {
+	readonly type: "insert";
+}
+
+/**
+ * {@link OperationStamp} for a 'set remove' operation. This aligns with the `mapRangeRemoved` API in MergeTree.
+ *
+ * @remarks The terminology here comes from the fact that the removal should affect only the *set* of nodes that were
+ * specified at the time the local client issued the remove, and not any nodes that were inserted concurrently.
+ *
+ * Not using "remove" and "obliterate" here allows us to unambiguously use the term "remove" elsewhere in code to mean
+ * "removed from the tree, either by MergeTree.obliterateRange or MergeTree.removeRange". This is convenient as the vast majority
+ * of merge-tree code only cares about segment visibility and not the specific operation that caused a segment to be removed.
+ */
+export interface SetRemoveOperationStamp extends OperationStamp {
+	readonly type: "setRemove";
+}
+
+/**
+ * {@link OperationStamp} for a 'set remove' operation. This aligns with the `mapRangeRemoved` API in MergeTree.
+ *
+ * @remarks The terminology here comes from the fact that the removal should affect the *slice* of nodes between the
+ * start and end point specified by the local client, which includes any nodes that were inserted concurrently.
+ *
+ * Not using "remove" and "obliterate" here allows us to unambiguously use the term "remove" elsewhere in code to mean
+ * "removed from the tree, either by MergeTree.obliterateRange or MergeTree.removeRange". This is convenient as the vast majority
+ * of merge-tree code only cares about segment visibility and not the specific operation that caused a segment to be removed.
+ */
+export interface SliceRemoveOperationStamp extends OperationStamp {
+	readonly type: "sliceRemove";
```

Comment:
i'm still figuring out how all of this is used. a bit concerned that obliterates are hard to follow, as we also conflate it with move. it is nice to be able to do a full text search and get most of the relevant bits. this is more about naming, as i do like reducing the need to differentiate them in code

Reply:
The `type` bit isn't really very important to the overall PR, and we can change it with ease if we want. Terminology-wise, I'd like to just be able to refer to a segment which is "gone from the tree" as "removed", since basically everywhere in merge-tree except specifically where we're dealing with obliterates, we treat segments that were removed by `markRangeRemoved` as identical to those that were removed by `markRangeObliterated`. Tree folks have been using set vs slice semantics to refer to the distinction between those two, so I figured I'd adopt that name.

Reply:
its really just about consistentcy. left a couple other comments, but will stop enumerating for now.